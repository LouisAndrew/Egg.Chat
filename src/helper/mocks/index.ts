export { mockChatroom } from './chatroom';
export { mockMessage } from './message';
export { mockUser2, mockUser1 } from './user';
export { mockFirebase } from './firebase';
