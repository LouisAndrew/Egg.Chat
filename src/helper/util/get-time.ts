const months = [
    'Jan',
    'Feb',
    'Mar',
    'Apr',
    'May',
    'Jun',
    'Jul',
    'Aug',
    'Sep',
    'Oct',
    'Nov',
    'Dec',
];

const getTime = (date: Date) => {
    const pad = (num: number) => (num < 10 ? `0${num}` : num);
    const dateAndMonth = (dt: Date) =>
        `${pad(date.getDate())} ${months[date.getMonth()]}`;
    const now = new Date();

    if (date.getMonth() !== now.getMonth()) {
        return dateAndMonth(date);
    }

    if (date.getDate() !== now.getDate()) {
        // create date object of yesterday timestamp.
        const yesterdayDate = new Date();
        yesterdayDate.setDate(yesterdayDate.getDate() - 1);

        if (yesterdayDate === date) {
            return 'Yesterday';
        } else {
            return dateAndMonth(date);
        }
    }

    return `${pad(date.getHours())}:${pad(date.getMinutes())}`;
};

export { getTime };
