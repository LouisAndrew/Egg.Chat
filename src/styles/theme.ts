import buttons from './buttons';
import text from './texts';
import variants from './variants';

const theme = {
    breakpoints: ['40em', '52em', '64em'],
    fontSizes: [12, 14, 16, 20, 24, 32, 48, 64],
    colors: {
        blue: {
            //       third,    blue-shade, blue-darker, blue-even-darker
            dark: ['#545E75', '#4B556C', '#3A4765', '#2F3A54'],
            //      primary
            light: '#CEEDFF',
        },
        //       black-shade, text-color
        black: ['#A5A5A5', '#333333'],
        //       secondary,  bg
        white: ['#F3F3F3', '#F7F7FF'],
        //      banner-new, banner-online
        misc: ['#FF8590', '#ABFFB3'],
    },
    space: [0, 4, 8, 16, 32, 64, 128, 256],
    fonts: {
        body: "'Open Sans', sans-serif",
        heading: "'Raleway', sans-serif",
        monospace: 'Menlo, monospace',
    },
    fontWeights: {
        body: 400,
        heading: 700,
        bold: 700,
    },
    lineHeights: {
        body: 1.5,
        heading: 1.25,
    },
    shadows: {
        small: '0 0 4px rgba(0, 0, 0, .125)',
        large: '0 0 24px rgba(0, 0, 0, .125)',
    },
    variants,
    text,
    buttons,
};

export default theme;
