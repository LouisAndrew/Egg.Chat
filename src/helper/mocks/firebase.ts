// use just as reference.
// Error message: Jest mock must be an inline function

import { Chatroom, User } from '../schema';
import { mockUser1, mockUser2, mockChatroom } from '.';

const mockUserDb: { [key: string]: User } = {
    '1234': mockUser1,
    '5678': mockUser2,
};

const mockChatroomDb: { [key: string]: Chatroom } = {
    '0972': mockChatroom,
};

export const mockFirebase = () => ({
    firestore: () => ({
        collection: (ref: string) => {
            if (ref === 'user') {
                return {
                    doc: (id: string) => ({
                        get: () =>
                            new Promise((res) => {
                                res(mockUserDb[id]);
                            }),
                    }),
                };
            } else if (ref === 'chatroom') {
                return {
                    doc: (id: string) => ({
                        get: () =>
                            new Promise((res) => {
                                res(mockChatroomDb[id]);
                            }),
                    }),
                };
            } else {
                return;
            }
        },
    }),
    auth: () => ({}),
});
