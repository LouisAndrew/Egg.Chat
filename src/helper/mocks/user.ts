import { Status, User } from '../schema';

const CHATROOM_ID = '0972';

const mockUser1: User = {
    uid: '1234',
    displayImage: '',
    displayName: 'John Doe',
    status: Status.ONLINE,
    chatrooms: [CHATROOM_ID],
    lastOnline: new Date(16023160238),
};

const mockUser2: User = {
    uid: '5678',
    displayImage: '',
    displayName: 'Jane Doe',
    status: Status.OFFLINE,
    chatrooms: [CHATROOM_ID],
    lastOnline: new Date(1605460238),
};

export { mockUser2, mockUser1 };
