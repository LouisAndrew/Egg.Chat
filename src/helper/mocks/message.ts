import { Message } from '../schema';
import { mockUser1 } from './user';

const mockMessage: Message = {
    msgId: '45725',
    msg: 'Hello, World',
    sentBy: mockUser1.uid,
    sentAt: new Date(Date.parse('10 Oct 2020 00:12:00 GMT')),
};

export { mockMessage };
