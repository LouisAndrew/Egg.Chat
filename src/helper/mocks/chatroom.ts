import { Chatroom } from 'helper/schema';
import { mockUser2, mockUser1 } from './user';
import { mockMessage } from './message';

const mockChatroom: Chatroom = {
    roomId: '0972',
    users: [mockUser1.uid, mockUser2.uid],
    messages: [mockMessage],
};

export { mockChatroom };
