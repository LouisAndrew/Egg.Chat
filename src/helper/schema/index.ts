import { Chatroom } from './chatroom';
import { User, Status } from './user';
import { Message } from './message';

export { Status };
export type { Chatroom, User, Message };
