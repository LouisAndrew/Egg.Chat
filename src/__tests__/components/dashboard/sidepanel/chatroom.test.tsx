import React from 'react';
import ReactDOM from 'react-dom';
import renderer from 'react-test-renderer';

import { render, cleanup } from '@testing-library/react';
import userEvent from '@testing-library/user-event';
import '@testing-library/jest-dom';

import Chatroom from 'components/dashboard/sidepanel/chatroom';
import { mockChatroom } from 'helper/mocks';
import { Chatroom as RoomSchema } from 'helper/schema';
import { getTime } from 'helper/util/get-time';
import theme from 'styles/theme';
import WithTheme from 'helper/util/with-theme';

describe('Chatroom', () => {
    const roomName = 'John Doe';
    const imgUrl = './src/img';

    const lastMsg = mockChatroom.messages[mockChatroom.messages.length - 1];

    // extending the chatroom mock to provide more details!
    const mockChatroomExtended: RoomSchema = {
        ...mockChatroom,
        roomName,
        imgUrl,
    };

    const mockChatroomEmpty: RoomSchema = {
        ...mockChatroomExtended,
        messages: [],
    };

    const mockSetActiveChatroom = jest.fn((str) => {});

    const Element = (
        <WithTheme>
            <Chatroom
                {...mockChatroomExtended}
                isActive={false}
                setActiveChatRoom={mockSetActiveChatroom}
                data-testid="wrapper"
            />
        </WithTheme>
    );

    const ElementNewNotif = (
        <Chatroom
            {...mockChatroomExtended}
            isActive={false}
            setActiveChatRoom={mockSetActiveChatroom}
            isNewNotification={true}
            data-testid="wrapper"
        />
    );

    const ElementActive = (
        <WithTheme>
            <Chatroom
                {...mockChatroomExtended}
                isActive={true}
                setActiveChatRoom={mockSetActiveChatroom}
                data-testid="wrapper"
            />
        </WithTheme>
    );

    const ElementEmpty = (
        <Chatroom
            {...mockChatroomEmpty}
            isActive={false}
            setActiveChatRoom={mockSetActiveChatroom}
            data-testid="wrapper"
        />
    );

    afterEach(cleanup);

    it('renders without crashing', () => {
        const div = document.createElement('div');
        ReactDOM.render(Element, div);
    });

    it('Should render the name of the room and its image correctly', () => {
        const { queryByText, queryByRole } = render(Element);

        const roomNameEl = queryByText(roomName);
        const img = queryByRole('img');

        expect(roomNameEl).toBeInTheDocument();
        expect(img).toBeInTheDocument();
        expect((img as HTMLImageElement).src).toBe(
            imgUrl.replace('.', 'http://localhost')
        );
    });

    it('Should render the last message in the room and when its sent', () => {
        const { queryByTestId } = render(Element);

        // querying component by lastMsg's last send content
        const lastMsgEl = queryByTestId('last-msg');
        // querying component by lastMsg's last sent date
        const lastMsgDate = queryByTestId('last-msg-sent');

        expect(lastMsgEl).toBeInTheDocument();
        expect(lastMsgEl).toHaveTextContent(lastMsg.msg);

        expect(lastMsgDate).toBeInTheDocument();
        expect(lastMsgDate).toHaveTextContent(getTime(lastMsg.sentAt));
    });

    it("Should not render last message and when its sent if the room's message is empty", () => {
        const { queryByTestId } = render(ElementEmpty);

        // querying component by lastMsg's last send content
        const lastMsgEl = queryByTestId('last-msg');
        // querying component by lastMsg's last sent date
        const lastMsgDate = queryByTestId('last-msg-sent');

        expect(lastMsgEl).not.toBeInTheDocument();
        expect(lastMsgDate).not.toBeInTheDocument();
    });

    it('Should not render the new notification badge if isNewNotification is not set', () => {
        const { queryByTestId } = render(Element);

        const newNotifBadge = queryByTestId('new-notification-badge');

        expect(newNotifBadge).not.toBeInTheDocument();
    });

    it('Should render the new notification badge if isNewNotification is set to true', () => {
        const { queryByTestId } = render(ElementNewNotif);

        const newNotifBadge = queryByTestId('new-notification-badge');

        expect(newNotifBadge).toBeInTheDocument();
    });

    it('Should not render the ACTIVE variant of the component if the room is currently active', () => {
        const { getByTestId } = render(Element);

        const wrapper = getByTestId('wrapper');

        expect(wrapper).not.toHaveStyle(
            `background-color: ${theme.colors.blue.dark[2]}`
        );
    });

    it('Should render the ACTIVE variant of the component if the room is currently active', () => {
        const { getByTestId } = render(ElementActive);

        const wrapper = getByTestId('wrapper');

        expect(wrapper).toHaveStyle(
            `background-color: ${theme.colors.blue.dark[2]}`
        );
    });

    it('Should call the setActiveChatRoom function if the compopnent is clicked', () => {
        const { getByTestId } = render(ElementActive);

        const wrapper = getByTestId('wrapper');

        userEvent.click(wrapper);

        expect(mockSetActiveChatroom).toBeCalled();
        expect(mockSetActiveChatroom).toBeCalledWith(mockChatroom.roomId);
    });

    it('matches snapshot', () => {
        const run = true;

        if (run) {
            const tree = renderer.create(Element).toJSON();
            expect(tree).toMatchSnapshot();
        }
    });
});
