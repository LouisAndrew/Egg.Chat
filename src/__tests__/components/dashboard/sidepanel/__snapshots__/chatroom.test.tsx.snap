// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Chatroom matches snapshot 1`] = `
<div
  className="css-yascl1"
  data-testid="wrapper"
  onClick={[Function]}
>
  <img
    className="css-1umo6ez"
    src="./src/img"
  />
  <div
    className="css-uq47rj"
  >
    <h2
      className="css-o4i2m7"
    >
      John Doe
    </h2>
    <div
      className="css-12f8mwu"
      data-testid="last-msg"
    >
      Hello, World
    </div>
  </div>
  <div
    className="css-6rtqbg"
    data-testid="last-msg-sent"
  >
    02:12
  </div>
</div>
`;
